src/Installer.php

```diff
@@ -160,8 +170,31 @@ private function installPuli()
 
         $httpClient = new HttpClient($this->disableTls, $this->cafile);
 
+        $versionUrl = sprintf(
+            static::VERSION_API_URL,
+            $this->disableTls ? 'http' : 'https'
+        );
+
+        for ($retries = 3; $retries > 0; --$retries) {
+            if (!($versions = json_decode($httpClient->download($versionUrl), true))) {
+                continue;
+            }
+
+            break;
+        }
+
+        if (0 === $retries) {
+            $this->error('The download failed repeatedly, aborting.');
+
+            return 1;
+        }
+
+        if (null === ($this->version = $this->matchVersion($versions))) {
```

Comment:
This current approach is a bit confusing IMO. I would introduce an if-block with three different methods:

```php
if (false !==  $this->version) {
    $this->validateVersion($this->version, $versions);
} elseif ('stable' === $this->stability) {
    $this->version = $this->getLatestStableVersion($versions);
} else {
    $this->version = $this->getLatestUnstableVersion($versions);
}
```

The `$stability` field doesn't exist yet. The field should contain "unstable" by default and overridden through the options `--stable` and `--unstable`. Once the first stable release is available, we'll change the default value to "stable".